File: android/lib/services/api_client.dart
import 'dart:typed_data';

import 'package:dio/dio.dart';

import '../models/save_slot.dart';

class ConflictException implements Exception {
  final int serverLastModifiedMs;
  final int clientLastModifiedMs;
  ConflictException(this.serverLastModifiedMs, this.clientLastModifiedMs);
}

class ApiClient {
  final Dio _dio;
  final String pin;

  ApiClient({required String baseUrl, required this.pin})
      : _dio = Dio(
          BaseOptions(
            baseUrl: baseUrl,
            connectTimeout: const Duration(seconds: 5),
            receiveTimeout: const Duration(seconds: 30),
            headers: {'x-sync-pin': pin},
          ),
        );

  Future<Map<String, dynamic>> health() async {
    final resp = await Dio().get('${_dio.options.baseUrl}/health');
    return resp.data as Map<String, dynamic>;
  }

  Future<List<SaveSlot>> listSaves() async {
    final resp = await _dio.get('/api/v1/saves');
    final slots = (resp.data['slots'] as List)
        .map((e) => SaveSlot.fromJson(e as Map<String, dynamic>))
        .toList();
    return slots;
  }

  /// Downloads a save slot and returns the raw ZIP bytes + server timestamp.
  Future<({Uint8List zipBytes, int serverLastModifiedMs})> downloadSave(
      String slotId) async {
    final resp = await _dio.get(
      '/api/v1/saves/$slotId/download',
      options: Options(responseType: ResponseType.bytes),
    );
    final serverMs = int.tryParse(
            resp.headers.value('x-slot-last-modified-ms') ?? '') ??
        0;
    return (
      zipBytes: Uint8List.fromList(resp.data as List<int>),
      serverLastModifiedMs: serverMs,
    );
  }

  /// Uploads ZIP bytes for [slotId].
  ///
  /// Throws [ConflictException] if the server copy is newer and [force] is false.
  Future<void> uploadSave(
    String slotId,
    Uint8List zipBytes,
    int clientLastModifiedMs, {
    bool force = false,
  }) async {
    try {
      await _dio.post(
        '/api/v1/saves/$slotId/upload${force ? '?force=true' : ''}',
        data: zipBytes,
        options: Options(
          headers: {
            'content-type': 'application/zip',
            'x-client-last-modified-ms': clientLastModifiedMs.toString(),
          },
        ),
      );
    } on DioException catch (e) {
      if (e.response?.statusCode == 409) {
        final body = e.response?.data as Map<String, dynamic>?;
        throw ConflictException(
          (body?['server_last_modified_ms'] as num?)?.toInt() ?? 0,
          clientLastModifiedMs,
        );
      }
      rethrow;
    }
  }

  Future<void> deleteSlot(String slotId) async {
    await _dio.delete('/api/v1/saves/$slotId');
  }
}

File: android/app/src/main/kotlin/com/stardewsync/data/api/ApiClient.kt
package com.stardewsync.data.api

import com.stardewsync.data.model.SaveSlotInfo
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.IOException
import java.util.concurrent.TimeUnit

class ApiClient(private val baseUrl: String, private val pin: String) {

    private val client = OkHttpClient.Builder()
        .connectTimeout(5, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    private fun Request.Builder.withPin() = header("x-sync-pin", pin)

    suspend fun health(): JSONObject = withContext(Dispatchers.IO) {
        val request = Request.Builder().url("$baseUrl/health").get().build()
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) throw IOException("Health check failed: ${response.code}")
            JSONObject(response.body.string())
        }
    }

    suspend fun listSaves(): List<SaveSlotInfo> = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("$baseUrl/api/v1/saves")
            .withPin()
            .get()
            .build()
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) throw IOException("List saves failed: ${response.code}")
            val array = JSONObject(response.body.string()).getJSONArray("slots")
            (0 until array.length()).map { SaveSlotInfo.fromJson(array.getJSONObject(it)) }
        }
    }

    // Returns Pair(zipBytes, serverLastModifiedMs)
    suspend fun downloadSave(slotId: String): Pair<ByteArray, Long> = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("$baseUrl/api/v1/saves/$slotId/download")
            .withPin()
            .get()
            .build()
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) throw IOException("Download failed: ${response.code}")
            val ts = response.header("x-slot-last-modified-ms")?.toLongOrNull() ?: 0L
            Pair(response.body.bytes(), ts)
        }
    }

    suspend fun uploadSave(
        slotId: String,
        zipBytes: ByteArray,
        clientLastModifiedMs: Long,
        force: Boolean = false,
    ) = withContext(Dispatchers.IO) {
        val url = "$baseUrl/api/v1/saves/$slotId/upload" + if (force) "?force=true" else ""
        val body = zipBytes.toRequestBody("application/octet-stream".toMediaType())
        val request = Request.Builder()
            .url(url)
            .withPin()
            .header("x-client-last-modified-ms", clientLastModifiedMs.toString())
            .post(body)
            .build()
        client.newCall(request).execute().use { response ->
            if (response.code == 409) {
                val json = JSONObject(response.body.string())
                throw ConflictException(
                    serverLastModifiedMs = json.getLong("server_last_modified_ms"),
                    clientLastModifiedMs = json.optLong("client_last_modified_ms", clientLastModifiedMs),
                )
            }
            if (!response.isSuccessful) throw IOException("Upload failed: ${response.code}")
        }
    }

    suspend fun deleteSlot(slotId: String) = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("$baseUrl/api/v1/saves/$slotId")
            .withPin()
            .delete()
            .build()
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) throw IOException("Delete failed: ${response.code}")
        }
    }

    suspend fun listServerBackups(): List<BackupInfo> = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("$baseUrl/api/v1/backups")
            .withPin()
            .get()
            .build()
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) throw IOException("List backups failed: ${response.code}")
            val array = org.json.JSONObject(response.body.string()).getJSONArray("backups")
            (0 until array.length()).map {
                val obj = array.getJSONObject(it)
                BackupInfo(
                    name = obj.getString("name"),
                    slotId = obj.getString("slot_id"),
                    timestampMs = obj.getLong("timestamp_ms"),
                )
            }
        }
    }

    suspend fun deleteServerBackup(name: String) = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("$baseUrl/api/v1/backups/$name")
            .withPin()
            .delete()
            .build()
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) throw IOException("Delete backup failed: ${response.code}")
        }
    }
}

data class BackupInfo(val name: String, val slotId: String, val timestampMs: Long)
